graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,194 @@
+    if len(filter_arguments) != 1:
+        raise AssertionError("Expected one argument in filter {}".format(filter_directive))
+
+    argument_name = cast(StringValueNode, filter_arguments[0]).value
+    parameter_name = get_parameter_name(argument_name)
+    return parameter_name
+
+
+def _get_filter_node_operation(filter_directive: DirectiveNode) -> str:
+    """Return the @filter's op_name as a string."""
+    return cast(StringValueNode, filter_directive.arguments[0].value).value
+
+
+def _add_pagination_filter(
+    query_ast: DocumentNode,
+    query_path: Tuple[str, ...],
+    pagination_field: str,
+    directive_to_add: DirectiveNode,
+) -> Tuple[DocumentNode, List[str]]:
+    """Add the filter to the target field, returning a query and the names of removed filters.
+
+    Args:
+        query_ast: The query in which we are adding a filter
+        query_path: The path to the pagination vertex
+        pagination_field: The field on which we are adding a filter
+        directive_to_add: The filter directive to add
+
+    Returns:
+        new_ast: A query with the filter inserted, and any filters on the same location with
+                 the same operation removed.
+        removed_parameters: The parameter names used in filters that were removed.
+    """
+    if not isinstance(query_ast, (FieldNode, InlineFragmentNode, OperationDefinitionNode)):
+        raise AssertionError(
+            f'Input AST is of type "{type(query_ast).__name__}", which should not be a selection.'
+        )
+
+    removed_parameters = []
+    new_selections = []
+    if len(query_path) == 0:
+        # Add the filter to the correct field on this vertex and remove any redundant filters.
+        found_field = False
+        for selection_ast in query_ast.selection_set.selections:
+            new_selection_ast = selection_ast
+            field_name = get_ast_field_name(selection_ast)
+            if field_name == pagination_field:
+                found_field = True
+                new_selection_ast = copy(selection_ast)
+                new_selection_ast.directives = copy(selection_ast.directives)
+
+                new_directives = []
+                for directive in selection_ast.directives:
+                    operation = _get_filter_node_operation(directive)
+                    if operation == _get_filter_node_operation(directive_to_add):
+                        removed_parameters.append(_get_binary_filter_node_parameter(directive))
+                    else:
+                        new_directives.append(directive)
+                new_directives.append(directive_to_add)
+                new_selection_ast.directives = new_directives
+            new_selections.append(new_selection_ast)
+        if not found_field:
+            new_selections.insert(
+                0, FieldNode(name=NameNode(value=pagination_field), directives=[directive_to_add])
+            )
+    else:
```

Comment:
Do you think there are potential refactorings that might make this function a bit easier on the eyes? There's a lot going on here, and it's not easy to follow.

Reply:
yes, working on it